### ISSUE TITLE
inconsistent next execution time in scheduler

### ISSUE THREAD

**Issue by 4d4mu** [smarthomeNG/smarthome#179]
Scheduler in backend plugin shows next execution time in:
- local time when `cycle` used
- local time when basic `crontab` used
- UTC when `crontab` with `sunrise`/`sunset` used

I'm not sure if it is backend or scheduler issue

**Comment by bmxp** [comment]
Can you please provide a sample? And do you refer to master branch oder are you running on develop?

**Comment by 4d4mu** [comment]
I'm on develop. Here's example (from Scheduler page):
```
next execution 		 	Cycle		Crontab
2017-01-09 04:52:00+0100			{'52 4 * *': None}
2017-01-09 07:02:00+0100			{'2 7 * 0,1,2,3,4': '1', '2 8 * 5,6': '1'}
2017-01-08 09:21:19+0100 	{300: None}
2017-01-08 15:40:24+0000			{'sunset-6': '1'}
2017-01-09 06:09:46+0000			{'6:40<sunrise-5': '1'}
```

**Comment by msinn** [comment]
@4d4mu I checked my schedules with the actual develop branch and don't have the problen (any more). Does the problem still exist on your side?

**Comment by bmxp** [comment]
The problem still persists with the uzsu when **sunset** or **sunrise** are used.

This is wrong
uzsu_UZSU Jalousien Westseite schliessen ----> 2017-10-16 16:26:58+0000 
![grafik](https://user-images.githubusercontent.com/3002409/31617001-835102fe-b28e-11e7-89e0-bb4963c0dc66.png)

This is correct
uzsu_UZSU Jalousien Westseite öffnen ---> 2017-10-17 07:20:00+0200
![grafik](https://user-images.githubusercontent.com/3002409/31617061-aad0e43e-b28e-11e7-9c5c-16e522160458.png)

So at the moment we need to examine the uzsu where sunset and sunrise are examined on the time settings.

**Comment by msinn** [comment]
That may have the origin in the UZSU plugin. I know sometime in the past, the system schedulers (for the env items) used GMT. I think that has changed (since I have no schedulers in GMT any more).

I don't use UZSU. Maybe someone could look into the plugin. When it adds a scheduler it controls which timezone is used.

**Comment by bmxp** [comment]
When I am done with my other urgent task except from SHNG I will have a look at it.

**Comment by 4d4mu** [comment]
Problem still occurs, I can see both UTC and local time in scheduler tab in backend.
SHNG develop 9e6d384
plugins develop afe97d7

**Comment by ohinckel** [comment]
I guess it has something to do with the `datetime` objects created in `lib.Orb` (see [set](https://github.com/smarthomeNG/smarthome/blob/develop/lib/orb.py#L71) or [rise()](https://github.com/smarthomeNG/smarthome/blob/develop/lib/orb.py#L57) methods). There's always `UTC` used (e.g. `datetime.datetime.utcnow()`).

**Comment by bmxp** [comment]
I have a fix locally for the uzsu and I am testing it. If you want to try out:
```
        # see if sunset or sunrise are included in give tstr param
        dmin = None
        dmax = None
        if cron.startswith('sunrise'):
            next_time = uzsu_sun.rise(doff, moff, dt=dt)
            # time in next_time will be in utctime. So we need to adjust it
            if next_time.tzinfo == tzutc():
                next_time = next_time.astimezone(self._sh.tzinfo())
            else:
                self.logger.warning("next_time.tzinfo was not given as utc!") 
            self.logger.debug("next_time.tzinfo gives {}".format(next_time.tzinfo)) 
            self.logger.debug("Sunrise is included and calculated as {}".format(next_time)) 
        elif cron.startswith('sunset'):
            next_time = uzsu_sun.set(doff, moff, dt=dt)
            # time in next_time will be in utctime. So we need to adjust it
            if next_time.tzinfo == tzutc():
                next_time = next_time.astimezone(self._sh.tzinfo())
            else:
                self.logger.warning("next_time.tzinfo was not given as utc!") 
            self.logger.debug("Sunset is included and calculated as {}".format(next_time)) 
        else:
            self.logger.error('Wrong syntax: {0}. Should be [H:M<](sunrise|sunset)[+|-][offset][<H:M]'.format(tstr))
            return
```

**Comment by ohinckel** [comment]
Really sure it is a plugin problem? To me it seems like a scheduler issue since some of my logics are also listed in UTC when they use sunrise/sunset.

We should not start mixing timezones at all the places. We should use one and always convert the `datetime` objects to the requested timezone. IMHO we should rely on the `tz` setting in smarthome configuration file (if not given use local time or UTC). If set all dates listed or shown somewhere should be in the configured timezone.

**Comment by bmxp** [comment]
In this case we have a uzsu plugin problem. The time of sunrise and sunset is given back as utc based time. Since all scheduler times should be aware times. There is some code to check that in scheduler. Maybe you want to copy some of your logics to a gist so I can have a look at it?

**Comment by msinn** [comment]
The scheduler doesn‘t do anything with timezeones, but respecting them. When you querry the schedules, it gives back the timezone that was thrown at it when adding tge schedule.

**Comment by ohinckel** [comment]
Yes, it is not a scheduler thing, but the scheduler is just using the datetime object with UTC without converting them. Usually this is not a problem, but we should always convert the dates delivered to "external" services (e.g. the CLI, the VISU, maybe the calendar, etc.) in the configured timezone.

For lib/orb.py we could pass the timezone to the constructor or create an environment object containing all env-specific setting which can be used anywhere.

But I would suggest to convert the dates to the configured timezone at all places were datetime object leaves the application providing them to external systems. This conversion should be implemented by on method, e.g. the in lib/utils.py.

**Comment by bmxp** [comment]
Will there be any possibility in far away future when the local coordinates and time zone can be configured during runtime? If not, we might use it within constructor ..

**Comment by msinn** [comment]
@bmxp Is this fixed with the modifications that went into v1.4.1?

**Comment by bmxp** [comment]
Yes, I think this is fixed for now but would need some refactoring in future together with replacement of pyephem by a python only lib and some changes in lib/orb and removing doubled code in lib/scheduler and uzsu. I close it for now.
